### PR TITLE
Introducing JSON representation model with serialisation methods

### PR DESCRIPTION
**Pull request by robsimpson**
# Description

Introduces a default Java model representation of a JSON object, with serialisation and deserialisation based on the existing DefaultWriter and DefaultReader implementations.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

**Comment by Robsimpson** [comment]
@microsoft-github-policy-service agree

File: sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+import java.io.OutputStream;
+import java.io.StringWriter;
+import java.io.Writer;
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Class representing the JSON array type
+ */
+public class JsonArray extends JsonElement {
```
[the remark this answers is not in the document]
Comment (Robsimpson) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:16]: Done @alzimmermsft

File: sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+     * @throws IOException if the writer throws an exception
+     */
+    public Writer toWriter(Writer writer) throws IOException {
+        if(writer != null){
+            try (JsonWriter jsonWriter = JsonProviders.createWriter(writer)) {
+                serialize(jsonWriter);
+            }
+            return writer;
+        } else {
+            throw new NullPointerException();
```
[the remark this answers is not in the document]
Comment (Robsimpson) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:222]: Easy - I'll carry this pattern through to the other null checks

File: sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+        }
+        input = input.replace(":", ": ");
+        return input;
+    }
+
+    /**
+     * Takes a writer and uses it to serialize the JsonArray object
+     * @param writer the writer to use, which is then wrapped in a JsonWriter
+     * @return the writer
+     * @throws IOException if the writer throws an exception
```
[the remark this answers is not in the document]
Comment (Robsimpson) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:213]: Done

File: sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+
+    /**
+     * Builds the JsonArray from an opened JsonReader.
+     * JsonReader is passed to the nested JsonElements to recursively build.
+     *
+     * @param reader the JsonReader to build the JsonArray from.
+     * @throws IOException Thrown when build is aborted due to encountering an
+     * invalid JsonToken indicating a improperly formed JsonArray.
+     */
+    private void build(JsonReader reader) throws IOException {
```
[the remark this answers is not in the document]
Comment (Robsimpson) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:298]: @alzimmermsft do you mean situations with a nested array and the rest of the tokens are still to follow, or a situation where an outer array is closed, and there are still tokens remaining in the reader? I thought our testing covered case one but reviewing that this evening to be sure

[the remark this answers is not in the document]
Comment (alzimmermsft) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:263]: I think both cases will need to be covered

- `[array stuff]. more JSON fields`
- `[array stuff], and we're in another array]`

This ends up being an issue overall, is the expectation that the JsonElement types can be deserialized from within a larger JSON scope or do they need to consume the entire JSON stream.

I'm leaning towards supporting partial deserialization as that should implicitly handle reading the entire stream. Continue to use JsonArray as an example, all of the following should work:

- `[json is just an array]`
- `[we're reading an array that has more JSON afterwards], more JSON`
- `[we're reading a nested array], outer array stuff]`

[the remark this answers is not in the document]
Comment (Robsimpson) [reply sdk/core/azure-json/src/main/java/com/azure/json/JsonArray.java:263]: Got it - apologies for slow reply, been bouncing around a bit at work. Will take a look at a few cases over the next couple of days and put something together

**Review by alzimmermsft — APPROVED** [review]

**Comment by robsimpson** [comment]
@microsoft-github-policy-service agree



Rob S
Sent from my iPhone

On 10 Apr 2024, at 3:26 AM, microsoft-github-policy-service[bot] ***@***.***> wrote:

@microsoft-github-policy-service agree

**Comment by alzimmermsft** [comment]
Thank you so much @robsimpson, @JM16450251, @ZacharyKn1ght, and @bronsonskipper, for this contribution!

**Pull request merged**